### PR TITLE
[typescript] Add parameter to enrich maps with propertyNames and required properties

### PR DESCRIPTION
**Pull request by Sqh3rd**
Add additional parameter for the abstract typescript codegen to consider defined propertyNames and required properties, when generating maps/dictionaries.

Implements #19433 

Mention TS-People:
(@TiFu @taxpon @sebastianhaas @kenisteward @Vrolijkx @macjohnny @topce @akehir @petejohansonxo @amakhrov  @davidgamero @mkusaka)

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.6.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

File: samples/client/echo_api/typescript-axios/build/.openapi-generator/VERSION [hunk 1]
```diff
@@ -1 +1 @@
-7.11.0-SNAPSHOT
+7.9.0-SNAPSHOT
```
Comment (macjohnny) [inline samples/client/echo_api/typescript-axios/build/.openapi-generator/VERSION:1]: can you please rebase on the latest master an re-generate the samples?

File: modules/openapi-generator/src/test/java/org/openapitools/codegen/typescript/TypeScriptClientCodegenTest.java [hunk 2]
```diff
@@ -47,6 +47,21 @@ public void getTypeDeclarationTest() {
...
+        parentSchema.required(List.of("x", "y"));
+        Assert.assertEquals(codegen.getTypeDeclaration(parentSchema), "{ [key: string]: string; }");
+
+        parentSchema.propertyNames(new StringSchema()._enum(List.of("y", "z")));
+        Assert.assertEquals(codegen.getTypeDeclaration(parentSchema), "{ [key: string]: string; }");
+
+        codegen.additionalProperties().put("enrichedMaps", Boolean.TRUE);
+        codegen.processOpts();
+        parentSchema.propertyNames(new StringSchema());
+        Assert.assertEquals(codegen.getTypeDeclaration(parentSchema), "{ [key: string]: string; x: string; y: string; }");
```
Comment (macjohnny) [inline modules/openapi-generator/src/test/java/org/openapitools/codegen/typescript/TypeScriptClientCodegenTest.java:61]: how about adding newlines between the properties in the generated code?

**Review by macjohnny — COMMENTED** [review]
thanks for your contribution!

**Comment by macjohnny** [comment]
cc @joscha

**Comment by fravl** [comment]
@Sqh3rd are you still working on this feature?

**Comment by Sqh3rd** [comment]
> @Sqh3rd are you still working on this feature?

will take it up again a bit later today and try to finish it by the end of the week :)

**Comment by Sqh3rd** [comment]
So after taking way longer than I thought it would, I'm not sure about the original approach of using a parameter to define this behavior.

In order to make the changes work neatly with the templates, it's necessary to add propertyNames in IJsonSchemaValidationProperties, additionally I added the nameOnlyVars list to be able to get them in the template. (I'd really appreciate some feedback about that, not sure if the solution fits well into the codebase)

Having a parameter that either sets or clears the corresponding vars seems very intrusive, so I'm not sure whether that makes sense to do. I think that just adding this behavior is not breaking backwards compatibility, as the properties will be optional, so I think it's fine.

**Comment by fravl** [comment]
@TiFu @taxpon @sebastianhaas @kenisteward @Vrolijkx @macjohnny @topce @akehir @petejohansonxo @amakhrov  @davidgamero @mkusaka @joscha

Tagging again, cause the PR is quite old. Could somebody from the technical committee please take a look once more at @Sqh3rd's work?

**Pull request closed without merging**